File: entities/hitbox.py
"""Hitbox / Hurtbox — separated damage and vulnerability regions.

Design
------
* **Hurtbox** — the fighter's vulnerable area. Always exists while alive.
  Supports per-state shrink/grow and crouching offsets.
* **Hitbox** — a transient damage region spawned by attacks. Tracks which
  ports it has already hit so each swing connects at most once per target.
* **HitboxGroup** — holds ≥1 hitboxes for a single attack, so a move can
  have a sweetspot / sourspot or a multi-part swing.
* **CollisionHelper** — pure-function AABB and circle-vs-rect tests used by
  both hitbox and projectile collision.
"""

from __future__ import annotations
import math
from dataclasses import dataclass, field
from enum import Enum, auto
from typing import List, Set, Tuple, TYPE_CHECKING

import pygame

if TYPE_CHECKING:
    from entities.fighter import Fighter


# ======================================================================
#  Shape helpers
# ======================================================================
class HitboxShape(Enum):
    RECT = auto()
    CIRCLE = auto()


# ======================================================================
#  Collision utilities
# ======================================================================
class CollisionHelper:
    """Pure-function static collision tests."""

    @staticmethod
    def rect_vs_rect(a: pygame.Rect, b: pygame.Rect) -> bool:
        return a.colliderect(b)

    @staticmethod
    def circle_vs_rect(cx: float, cy: float, r: float,
                       rect: pygame.Rect) -> bool:
        """Test circle (cx, cy, r) against axis-aligned rect."""
        # Closest point on rect to circle center
        closest_x = max(rect.left, min(cx, rect.right))
        closest_y = max(rect.top, min(cy, rect.bottom))
        dx = cx - closest_x
        dy = cy - closest_y
        return (dx * dx + dy * dy) <= r * r

    @staticmethod
    def rect_vs_platforms(rect: pygame.Rect, platforms) -> bool:
        """Return True if *rect* overlaps any solid (non-passthrough) platform."""
        for plat in platforms:
            if not plat.is_passthrough and rect.colliderect(plat.rect):
                return True
        return False


# ======================================================================
#  Hurtbox
# ======================================================================
@dataclass
class HurtboxProfile:
    """Per-state size overrides (relative to FighterData width/height)."""
    width_scale: float = 1.0
    height_scale: float = 1.0
    offset_y: float = 0.0           # shift down (+) for crouch, up (-) for jump


# Default profiles for common states
DEFAULT_HURTBOX_PROFILES = {
    "idle":       HurtboxProfile(1.0,  1.0,    0),
    "walking":    HurtboxProfile(1.0,  1.0,    0),
    "running":    HurtboxProfile(1.0,  0.9,    4),
    "crouching":  HurtboxProfile(1.1,  0.65,  14),
    "airborne":   HurtboxProfile(0.9,  0.95,   0),
    "hitstun":    HurtboxProfile(1.0,  1.0,    0),
    "shielding":  HurtboxProfile(0.85, 0.85,   4),
}


class Hurtbox:
    """A fighter's vulnerable collision area.

    The hurtbox is always centred on the fighter's position and sized
    according to the current state profile.
    """

    def __init__(self, fighter: Fighter):
        self.fighter = fighter
        self.profiles = dict(DEFAULT_HURTBOX_PROFILES)

    def get_rect(self, state_name: str | None = None) -> pygame.Rect:
        """Return world-space hurtbox rect for the fighters current state."""
        key = state_name or self.fighter.state.name.lower()
        prof = self.profiles.get(key, HurtboxProfile())
        w = int(self.fighter.data.width * prof.width_scale)
        h = int(self.fighter.data.height * prof.height_scale)
        x = int(self.fighter.x + (self.fighter.data.width - w) / 2)
        y = int(self.fighter.y + (self.fighter.data.height - h) + prof.offset_y)
        return pygame.Rect(x, y, w, h)


# ======================================================================
#  Hitbox
# ======================================================================
class Hitbox:
    """A transient damage region that lives for a set number of active frames.

    Attributes
    ----------
    shape : HitboxShape
        RECT or CIRCLE.  When CIRCLE, *width* is treated as diameter and
        *height* is ignored.
    priority : int
        Higher-priority hitbox wins a hitbox-vs-hitbox trade.
    """

    def __init__(
        self,
        owner_port: int,
        offset_x: float,
        offset_y: float,
        width: float,
        height: float,
        damage: float,
        base_knockback: float,
        kb_scaling: float,
        angle: float,
        active_frames: int,
        shape: HitboxShape = HitboxShape.RECT,
        priority: int = 0,
    ):
        self.owner_port = owner_port
        self.offset_x = offset_x
        self.offset_y = offset_y
        self.width = width
        self.height = height
        self.damage = damage
        self.base_knockback = base_knockback
        self.kb_scaling = kb_scaling
        self.angle = angle              # degrees: 0=right, 90=up, 270=down
        self.shape = shape
        self.priority = priority

        # Active frames bookkeeping
        self.total_active_frames = active_frames
        self.frames_remaining = active_frames
        self.already_hit: Set[int] = set()

    # ----- query ---
    def is_active(self) -> bool:
        return self.frames_remaining > 0

    @property
    def elapsed_frames(self) -> int:
        return self.total_active_frames - self.frames_remaining

    def tick(self) -> None:
        if self.frames_remaining > 0:
            self.frames_remaining -= 1

    # ----- geometry ---
    def get_world_rect(self, anchor_x: float, anchor_y: float,
                       anchor_w: float, anchor_h: float,
                       facing: int) -> pygame.Rect:
        """Compute world-space AABB (bounding box for circle shapes too)."""
        cx = anchor_x + anchor_w / 2
        cy = anchor_y + anchor_h / 2
        ox = self.offset_x * facing
        if self.shape == HitboxShape.CIRCLE:
            r = self.width / 2
            return pygame.Rect(int(cx + ox - r), int(cy + self.offset_y - r),
                               int(self.width), int(self.width))
        x = cx + ox - self.width / 2
        y = cy + self.offset_y - self.height / 2
        return pygame.Rect(int(x), int(y), int(self.width), int(self.height))

    def get_world_center(self, anchor_x: float, anchor_y: float,
                         anchor_w: float, anchor_h: float,
                         facing: int) -> Tuple[float, float]:
        cx = anchor_x + anchor_w / 2 + self.offset_x * facing
        cy = anchor_y + anchor_h / 2 + self.offset_y
        return cx, cy

    # ----- launch direction ---
    def get_launch_angle(self, facing: int) -> float:
        if facing == -1:
            return 180 - self.angle
        return self.angle

    # ----- hit tests ---
    def _overlaps(self, anchor_x: float, anchor_y: float,
                  anchor_w: float, anchor_h: float,
                  facing: int, target_rect: pygame.Rect) -> bool:
        """Shape-aware overlap test against an axis-aligned target rect."""
        if self.shape == HitboxShape.CIRCLE:
            cx, cy = self.get_world_center(
                anchor_x, anchor_y, anchor_w, anchor_h, facing)
            return CollisionHelper.circle_vs_rect(
                cx, cy, self.width / 2, target_rect)
        world_rect = self.get_world_rect(
            anchor_x, anchor_y, anchor_w, anchor_h, facing)
        return CollisionHelper.rect_vs_rect(world_rect, target_rect)

    def check_hit(self, attacker: Fighter, target: Fighter) -> bool:
        """Test collision against *target*'s hurtbox. Returns True on new hit."""
        if target.port in self.already_hit:
            return False
        if target.invincible:
            return False
        if target.port == self.owner_port:
            return False

        target_hurtbox = target.get_hurtbox()
        if self._overlaps(attacker.x, attacker.y,
                          attacker.width, attacker.height,
                          attacker.facing, target_hurtbox):
            self.already_hit.add(target.port)
            return True
        return False

    def check_hit_at(self, wx: float, wy: float, w: float, h: float,
                     facing: int, target: Fighter) -> bool:
        """Test from arbitrary world position (projectiles)."""
        if target.port in self.already_hit:
            return False
        if target.invincible:
            return False
        if target.port == self.owner_port:
            return False

        target_hurtbox = target.get_hurtbox()
        if self._overlaps(wx, wy, w, h, facing, target_hurtbox):
            self.already_hit.add(target.port)
            return True
        return False

    # ----- factory ---
    @staticmethod
    def from_attack_data(owner_port: int, attack_data) -> Hitbox:
        """Create a Hitbox from an AttackData instance."""
        shape = HitboxShape.RECT
        if hasattr(attack_data, "hitbox_shape"):
            shape = {
                "circle": HitboxShape.CIRCLE,
                "rect": HitboxShape.RECT,
            }.get(attack_data.hitbox_shape, HitboxShape.RECT)

        priority = getattr(attack_data, "hitbox_priority", 0)

        return Hitbox(
            owner_port=owner_port,
            offset_x=attack_data.hitbox_x,
            offset_y=attack_data.hitbox_y,
            width=attack_data.hitbox_w,
            height=attack_data.hitbox_h,
            damage=attack_data.damage,
            base_knockback=attack_data.base_knockback,
            kb_scaling=attack_data.kb_scaling,
            angle=attack_data.angle,
            active_frames=attack_data.active_frames,
            shape=shape,
            priority=priority,
        )


# ======================================================================
#  HitboxGroup  — multi-hitbox per attack (sweetspot / sourspot)
# ======================================================================
class HitboxGroup:
    """Container for ≥1 hitboxes that belong to the same attack swing.

    Only the first hitbox to connect on a given target applies (highest
    priority wins when multiple overlap the same hurtbox on the same frame).
    """

    def __init__(self, hitboxes: List[Hitbox]):
        self._hitboxes = sorted(hitboxes, key=lambda h: -h.priority)

    @property
    def hitboxes(self) -> List[Hitbox]:
        return self._hitboxes

    def is_active(self) -> bool:
        return any(h.is_active() for h in self._hitboxes)

    def tick(self) -> None:
        for h in self._hitboxes:
            h.tick()

    def check_hit(self, attacker: Fighter,
                  target: Fighter) -> Hitbox | None:
        """Return the highest-priority hitbox that connects, or None."""
        for hb in self._hitboxes:
            if not hb.is_active():
                continue
            if hb.check_hit(attacker, target):
                # Mark other hitboxes so target isn't hit twice
                for other in self._hitboxes:
                    other.already_hit.add(target.port)
                return hb
        return None

    @staticmethod
    def from_attack_data(owner_port: int, attack_data) -> HitboxGroup:
        """Build from AttackData.  If the attack defines *extra_hitboxes*
        (list of dicts) they become additional hitboxes; otherwise a
        single-hitbox group is created."""
        primary = Hitbox.from_attack_data(owner_port, attack_data)
        extras: List[Hitbox] = []

        if hasattr(attack_data, "extra_hitboxes"):
            for ex in (attack_data.extra_hitboxes or []):
                extras.append(Hitbox(
                    owner_port=owner_port,
                    offset_x=ex.get("x", 0),
                    offset_y=ex.get("y", 0),
                    width=ex.get("w", 40),
                    height=ex.get("h", 40),
                    damage=ex.get("damage", primary.damage * 0.6),
                    base_knockback=ex.get("base_kb", primary.base_knockback * 0.5),
                    kb_scaling=ex.get("kb_scaling", primary.kb_scaling),
                    angle=ex.get("angle", primary.angle),
                    active_frames=ex.get("active_frames", primary.total_active_frames),
                    shape=HitboxShape.RECT,
                    priority=ex.get("priority", -1),
                ))

        return HitboxGroup([primary, *extras])

    def get_first_active(self) -> Hitbox | None:
        """Return the first still-active hitbox (for rendering / debug)."""
        for h in self._hitboxes:
            if h.is_active():
                return h
        return None

File: js/Hitbox.js
/**
 * Hitbox.js — Rect and circle hitbox shapes, hurtbox, collision math.
 */
(function() {
const S = SMASH.Settings;

// =====================================================================
// Collision utilities
// =====================================================================
const Collision = {
    /** AABB vs AABB */
    rectRect(ax, ay, aw, ah, bx, by, bw, bh) {
        return ax < bx + bw && ax + aw > bx && ay < by + bh && ay + ah > by;
    },

    /** Circle vs AABB */
    circleRect(cx, cy, cr, rx, ry, rw, rh) {
        const closestX = Math.max(rx, Math.min(cx, rx + rw));
        const closestY = Math.max(ry, Math.min(cy, ry + rh));
        const dx = cx - closestX;
        const dy = cy - closestY;
        return (dx * dx + dy * dy) <= (cr * cr);
    },

    /** Circle vs Circle */
    circleCircle(ax, ay, ar, bx, by, br) {
        const dx = ax - bx;
        const dy = ay - by;
        const sum = ar + br;
        return (dx * dx + dy * dy) <= (sum * sum);
    },
};

// =====================================================================
// Hurtbox — always a rect, separate from sprite bounds
// =====================================================================
class Hurtbox {
    constructor(fighter) {
        this.fighter = fighter;
        // Slightly narrower than visual for fairness
        this.widthRatio  = 0.75;
        this.heightRatio = 0.92;
    }

    getRect() {
        const f = this.fighter;
        const w = f.data.width  * this.widthRatio;
        const h = f.data.height * this.heightRatio;
        const x = f.x + (f.data.width  - w) / 2;
        const y = f.y + (f.data.height - h);
        return { x, y, w, h };
    }
}

// =====================================================================
// Hitbox — per-attack damage region (rect OR circle)
// =====================================================================
class Hitbox {
    /**
     * @param {object} opts
     * @param {'rect'|'circle'} opts.shape
     * @param {number} opts.ownerPort
     * @param {number} opts.offsetX      offset from fighter center (facing right)
     * @param {number} opts.offsetY      offset from fighter center
     * @param {number} [opts.width]      rect width
     * @param {number} [opts.height]     rect height
     * @param {number} [opts.radius]     circle radius
     * @param {number} opts.damage
     * @param {number} opts.baseKB
     * @param {number} opts.kbScaling
     * @param {number} opts.angle        launch angle (degrees, 0=right 90=up)
     * @param {number} opts.activeFrames
     */
    constructor(opts) {
        this.shape      = opts.shape || 'rect';
        this.ownerPort  = opts.ownerPort;
        this.offsetX    = opts.offsetX || 0;
        this.offsetY    = opts.offsetY || 0;
        this.width      = opts.width  || 60;
        this.height     = opts.height || 40;
        this.radius     = opts.radius || 30;
        this.damage     = opts.damage || 10;
        this.baseKB     = opts.baseKB || 200;
        this.kbScaling  = opts.kbScaling || 1.0;
        this.angle      = opts.angle  || 45;
        this.activeFrames    = opts.activeFrames || 3;
        this.framesRemaining = opts.activeFrames || 3;
        this.alreadyHit = new Set();
    }

    isActive() { return this.framesRemaining > 0; }

    tick() { if (this.framesRemaining > 0) this.framesRemaining--; }

    /** World-space center of the hitbox. */
    getWorldCenter(fx, fy, fw, fh, facing) {
        const cx = fx + fw / 2 + this.offsetX * facing;
        const cy = fy + fh / 2 + this.offsetY;
        return { x: cx, y: cy };
    }

    /** World-space bounding rect (for rect shape or bounding box of circle). */
    getWorldRect(fx, fy, fw, fh, facing) {
        const c = this.getWorldCenter(fx, fy, fw, fh, facing);
        if (this.shape === 'circle') {
            return {
                x: c.x - this.radius,
                y: c.y - this.radius,
                w: this.radius * 2,
                h: this.radius * 2,
            };
        }
        return {
            x: c.x - this.width / 2,
            y: c.y - this.height / 2,
            w: this.width,
            h: this.height,
        };
    }

    /** Launch angle adjusted for facing direction. */
    getLaunchAngle(facing) {
        return facing === -1 ? 180 - this.angle : this.angle;
    }

    /**
     * Test collision against a target fighter's hurtbox.
     * @returns {boolean} true on NEW hit
     */
    checkHit(attackerFighter, targetFighter) {
        if (targetFighter.port === this.ownerPort) return false;
        if (this.alreadyHit.has(targetFighter.port)) return false;
        if (targetFighter.invincible) return false;

        const hr = targetFighter.hurtbox.getRect();
        const c  = this.getWorldCenter(
            attackerFighter.x, attackerFighter.y,
            attackerFighter.data.width, attackerFighter.data.height,
            attackerFighter.facing
        );

        let hit = false;
        if (this.shape === 'circle') {
            hit = Collision.circleRect(c.x, c.y, this.radius, hr.x, hr.y, hr.w, hr.h);
        } else {
            const wr = this.getWorldRect(
                attackerFighter.x, attackerFighter.y,
                attackerFighter.data.width, attackerFighter.data.height,
                attackerFighter.facing
            );
            hit = Collision.rectRect(wr.x, wr.y, wr.w, wr.h, hr.x, hr.y, hr.w, hr.h);
        }

        if (hit) this.alreadyHit.add(targetFighter.port);
        return hit;
    }

    /**
     * Test collision from arbitrary position (for projectiles).
     */
    checkHitAt(wx, wy, pw, ph, facing, targetFighter) {
        if (targetFighter.port === this.ownerPort) return false;
        if (this.alreadyHit.has(targetFighter.port)) return false;
        if (targetFighter.invincible) return false;

        const hr = targetFighter.hurtbox.getRect();
        const c  = this.getWorldCenter(wx, wy, pw, ph, facing);

        let hit = false;
        if (this.shape === 'circle') {
            hit = Collision.circleRect(c.x, c.y, this.radius, hr.x, hr.y, hr.w, hr.h);
        } else {
            const wr = this.getWorldRect(wx, wy, pw, ph, facing);
            hit = Collision.rectRect(wr.x, wr.y, wr.w, wr.h, hr.x, hr.y, hr.w, hr.h);
        }

        if (hit) this.alreadyHit.add(targetFighter.port);
        return hit;
    }

    /** Projectile-vs-projectile collision (bounding circles). */
    checkProjectileCollision(selfPos, selfSize, otherPos, otherSize) {
        const r1 = Math.max(selfSize.w, selfSize.h) / 2;
        const r2 = Math.max(otherSize.w, otherSize.h) / 2;
        return Collision.circleCircle(
            selfPos.x + selfSize.w/2, selfPos.y + selfSize.h/2, r1,
            otherPos.x + otherSize.w/2, otherPos.y + otherSize.h/2, r2
        );
    }

    /** Create from AttackData. */
    static fromAttack(ownerPort, atk) {
        return new Hitbox({
            shape:        atk.hitboxShape || 'rect',
            ownerPort,
            offsetX:      atk.hitboxX,
            offsetY:      atk.hitboxY,
            width:        atk.hitboxW,
            height:       atk.hitboxH,
            radius:       atk.hitboxR || Math.max(atk.hitboxW, atk.hitboxH) / 2,
            damage:       atk.damage,
            baseKB:       atk.baseKB,
            kbScaling:    atk.kbScaling,
            angle:        atk.angle,
            activeFrames: atk.activeFrames,
        });
    }
}

// =====================================================================
// Debug rendering helpers
// =====================================================================
Hitbox.debugDraw = function(ctx, hitbox, fx, fy, fw, fh, facing, cam) {
    if (!S.DEBUG_HITBOXES) return;
    const c = hitbox.getWorldCenter(fx, fy, fw, fh, facing);
    const sx = cam.wtsx(c.x);
    const sy = cam.wtsy(c.y);

    ctx.save();
    ctx.globalAlpha = 0.35;
    if (hitbox.shape === 'circle') {
        ctx.fillStyle = '#ff3030';
        ctx.beginPath();
        ctx.arc(sx, sy, hitbox.radius * cam.zoom, 0, Math.PI * 2);
        ctx.fill();
        ctx.strokeStyle = '#ff0000';
        ctx.lineWidth = 2;
        ctx.stroke();
    } else {
        const wr = hitbox.getWorldRect(fx, fy, fw, fh, facing);
        ctx.fillStyle = '#ff3030';
        ctx.fillRect(
            cam.wtsx(wr.x), cam.wtsy(wr.y),
            wr.w * cam.zoom, wr.h * cam.zoom
        );
        ctx.strokeStyle = '#ff0000';
        ctx.lineWidth = 2;
        ctx.strokeRect(
            cam.wtsx(wr.x), cam.wtsy(wr.y),
            wr.w * cam.zoom, wr.h * cam.zoom
        );
    }
    ctx.restore();
};

Hurtbox.debugDraw = function(ctx, hurtbox, cam) {
    if (!S.DEBUG_HURTBOXES) return;
    const r = hurtbox.getRect();
    ctx.save();
    ctx.globalAlpha = 0.25;
    ctx.fillStyle = '#30ff30';
    ctx.fillRect(cam.wtsx(r.x), cam.wtsy(r.y), r.w * cam.zoom, r.h * cam.zoom);
    ctx.strokeStyle = '#00ff00';
    ctx.lineWidth = 1;
    ctx.strokeRect(cam.wtsx(r.x), cam.wtsy(r.y), r.w * cam.zoom, r.h * cam.zoom);
    ctx.restore();
};

// Export
SMASH.Collision = Collision;
SMASH.Hurtbox   = Hurtbox;
SMASH.Hitbox    = Hitbox;
})();
